src/feature/dirauth/process_descs.c

```diff
@@ -428,15 +428,20 @@ dirserv_free_fingerprint_list(void)
 
 /** Return -1 if <b>ri</b> has a private or otherwise bad address,
  * unless we're configured to not care. Return 0 if all ok. */
-static int
+int
 dirserv_router_has_valid_address(routerinfo_t *ri)
 {
   tor_addr_t addr;
   if (get_options()->DirAllowPrivateAddresses)
     return 0; /* whatever it is, we're fine with it */
   tor_addr_from_ipv4h(&addr, ri->addr);
 
-  if (tor_addr_is_internal(&addr, 0)) {
+  /* We only check internal v6 on non-null addresses because we do not require
+   * IPv6 and null IPv6 is normal. */
```

Comment:
This code would be easier to understand if you split the IPv4 and IPv6 cases into separate `if` statements.